File: cpp/medium/NumberofIslands.cpp
#include <cmath>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
  int directions[4][2] = {{0, -1}, {-1, 0}, {1, 0}, {0, 1}};

public:
  int numIslands(vector<vector<char>> &grid) {
    int ROWS = grid.size(), COLUM = grid[0].size();
    int numOfIslands = 0;
    for (int r = 0; r < ROWS; r++) {

      for (int c = 0; c < COLUM; c++) {
        if (grid[r][c] == '1') {
          DFS(grid, r, c);
          numOfIslands++;
        }
      }
    }
    return numOfIslands;
  }

  void DFS(vector<vector<char>> &grid, int r, int c) {
    if (r < 0 || c < 0 || r >= grid.size() || c >= grid[0].size() ||
        grid[r][c] == '0') {
      return;
    }

    grid[r][c] = '0';
    for (int i = 0; i < 4; i++) {
      DFS(grid, r + directions[i][0], c + directions[i][1]);
    }
  }
};

int main() {}

File: stonycodes70proplems/medium/NumberofIslands.py
from typing import List


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        DIRECTIONS = [(1, 0), (0, 1), (0, -1), (-1, 0)]

        def DFS(row: int, col: int):
            if (
                row < 0
                or row >= len(grid)
                or col < 0
                or col >= len(grid[0])
                or grid[row][col] == "0"
            ):
                return
            grid[row][col] = "0"
            for x, y in DIRECTIONS:
                DFS(row + x, col + y)

        ans = 0
        for row in range(len(grid)):
            for col in range(len(grid[0])):
                if grid[row][col] == "1":
                    DFS(row, col)
                    ans += 1

        return ans
